atomicfu-transformer/src/main/kotlin/kotlinx/atomicfu/transformer/AtomicFUTransformer.kt

```diff
@@ -785,33 +775,71 @@ class AtomicFUTransformer(
+            }
+            // start of array element operation arguments
+            val args = getter.next
+            // remove array element getter
+            instructions.remove(getter)
+            val arrayElementOperation = FlowAnalyzer(args).execute()
+            return fixupOperationOnAtomicVar(arrayElementOperation, f, ld, arrayElementInfo)
+        }
+
+        private fun fixupOperationOnAtomicVar(operation: AbstractInsnNode, f: FieldInfo, ld: FieldInsnNode, arrayElementInfo: List<AbstractInsnNode>?): AbstractInsnNode? {
+            val onArrayElement = f.getPrimitiveType(vh).sort == ARRAY
+            when (operation) {
                 is MethodInsnNode -> {
                     // invoked virtual method on atomic var -- fixup & done with it
-                    debug("invoke $f.${j.name}", sourceInfo.copy(i = j))
-                    return fixupInvokeVirtual(ld, false, j, f)
+                    debug("invoke $f.${operation.name}", sourceInfo.copy(i = operation))
+                    return fixupInvokeVirtual(ld, onArrayElement, operation, f)
                 }
                 is VarInsnNode -> {
                     // was stored to local -- needs more processing:
                     // store owner ref into the variable instead
-                    val v = j.`var`
-                    val next = j.next
-                    instructions.remove(ld)
-                    val lv = localVar(v, j)
+                    val v = operation.`var`
+                    val next = operation.next
+                    if (onArrayElement) {
+                        // leave just owner class load insn on stack
+                        arrayElementInfo!!.forEach { instructions.remove(it) }
+                    } else {
+                        instructions.remove(ld)
+                    }
+                    val lv = localVar(v, operation)
                     if (lv != null) {
                         // Stored to a local variable with an entry in LVT (typically because of inline function)
-                        if (lv.desc != f.fieldType.descriptor)
+                        if (lv.desc != f.fieldType.descriptor && !onArrayElement)
                             abort("field $f was stored to a local variable #$v \"${lv.name}\" with unexpected type: ${lv.desc}")
                         // correct local variable descriptor
                         lv.desc = f.ownerType.descriptor
                         lv.signature = null
                         // process all loads of this variable in the corresponding local variable range
                         forVarLoads(v, lv.start, lv.end) { otherLd ->
-                            fixupVarLoad(f, ld, otherLd)
+                            if (!onArrayElement)
+                                fixupVarLoad(f, ld, otherLd)
+                            else {
+                                fixupArrayElementLoad(f, ld, otherLd, arrayElementInfo!!)
+                            }
```

Comment:
This whole `if (!onArrayElement) { fixupVarLoad(...) } else {  fixupArrayElementLoad(...) }` is copied in two places of code. It is better to extract it into some kind of `fixupLoad` method. 

Also, style note: prefer writing `if (onArrayElement) ...` (direct form) as opposed to `if (! onArrayElement) ...` (negated). There is a refactoring to "invert if".